crates/core/src/network/connection_manager.rs

```diff
@@ -71,31 +73,148 @@ impl Merge for GenStatus {
 
 struct ConnectionManagerInner {
     router: Arc<MessageRouter>,
-    connection_by_gen_id: HashMap<GenerationalNodeId, Vec<Connection>>,
+    connection_by_gen_id: HashMap<(GenerationalNodeId, Swimlane), Connection>,
     /// This tracks the max generation we observed from connection attempts regardless of our nodes
     /// configuration. We cannot accept connections from nodes older than ones we have observed
     /// already.
     observed_generations: HashMap<PlainNodeId, GenStatus>,
 }
 
 impl ConnectionManagerInner {
-    fn get_random_connection(
+    fn get_connection(
         &self,
-        peer_node_id: &GenerationalNodeId,
-        target_concurrency: usize,
+        peer_node_id: GenerationalNodeId,
+        swimlane: Swimlane,
     ) -> Option<Connection> {
-        use rand::prelude::IndexedRandom;
         self.connection_by_gen_id
-            .get(peer_node_id)
-            .and_then(|connections| {
-                // Suggest we create new connection if the number
-                // of connections is below the target
-                if connections.len() >= target_concurrency {
-                    connections.choose(&mut rand::rng()).cloned()
-                } else {
-                    None
+            .get(&(peer_node_id, swimlane))
+            .cloned()
+    }
+
+    fn create_loopback_connection(
+        &mut self,
+        swimlane: Swimlane,
+        connection_manager: ConnectionManager,
+    ) -> Result<Connection, ConnectError> {
+        let (tx, egress, shared) = EgressStream::create_loopback();
+        let connection = Connection::new(
+            my_node_id(),
+            restate_types::net::CURRENT_PROTOCOL_VERSION,
+            swimlane,
+            tx,
+        );
+
+        let reactor = ConnectionReactor::new(connection.clone(), shared, None, self.router.clone());
+
+        let previous = self.register(connection.clone());
```

Comment:
Why is the explicit register call required here? It looks that it will also happen through the `ConnectionTracking` trait which is called in the `ConnectionReactor`.

Reply:
If we didn't do that, there could be more loopback connections than we need because we let go of the lock between these two steps.